File: clients/python/zeppelin/rank_by.py
"""RankBy builder for BM25 full-text search queries.

Usage::

    from zeppelin.rank_by import RankBy

    # Single field
    r = RankBy.bm25("content", "search query")

    # Multi-field with weights
    r = RankBy.sum(
        RankBy.product(2.0, RankBy.bm25("title", "query")),
        RankBy.bm25("body", "query"),
    )
"""

from __future__ import annotations

from typing import Any

RankByExpr = list[Any]


class RankBy:
    """Static methods that produce rank_by S-expression arrays."""

    @staticmethod
    def bm25(field: str, query: str) -> RankByExpr:
        """Single-field BM25: ``["field", "BM25", "query"]``"""
        return [field, "BM25", query]

    @staticmethod
    def sum(*exprs: RankByExpr) -> RankByExpr:
        """Sum of multiple expressions: ``["Sum", [...exprs]]``"""
        return ["Sum", list(exprs)]

    @staticmethod
    def max(*exprs: RankByExpr) -> RankByExpr:
        """Max of multiple expressions: ``["Max", [...exprs]]``"""
        return ["Max", list(exprs)]

    @staticmethod
    def product(weight: float, expr: RankByExpr) -> RankByExpr:
        """Weighted expression: ``["Product", weight, expr]``"""
        return ["Product", weight, expr]

File: clients/typescript/src/rank-by.ts
import type { RankByExpr } from "./types.js";

/**
 * RankBy builder with static methods that produce S-expression arrays
 * matching the Zeppelin wire format.
 *
 * @example
 * ```ts
 * import { RankBy } from "@zeppelin/client";
 *
 * // Single field
 * RankBy.bm25("content", "search query");
 *
 * // Weighted multi-field
 * RankBy.sum(
 *   RankBy.product(2.0, RankBy.bm25("title", "q")),
 *   RankBy.bm25("body", "q"),
 * );
 * ```
 */
export const RankBy = {
  /** Single-field BM25: `["field", "BM25", "query"]` */
  bm25(field: string, query: string): RankByExpr {
    return [field, "BM25", query];
  },

  /** Sum of multiple expressions: `["Sum", [...exprs]]` */
  sum(...exprs: RankByExpr[]): RankByExpr {
    return ["Sum", exprs];
  },

  /** Max of multiple expressions: `["Max", [...exprs]]` */
  max(...exprs: RankByExpr[]): RankByExpr {
    return ["Max", exprs];
  },

  /** Weighted expression: `["Product", weight, expr]` */
  product(weight: number, expr: RankByExpr): RankByExpr {
    return ["Product", weight, expr];
  },
} as const;
